sql/src/main/java/org/apache/druid/sql/calcite/rule/DruidFreeUnionDataSourceRule.java

```diff
@@ -0,0 +1,150 @@
+            operand(DruidRel.class, none()),
+            operand(DruidRel.class, none())
+        )
+    );
+    this.plannerContext = plannerContext;
+  }
+
+  @Override
+  public boolean matches(RelOptRuleCall call)
+  {
+    final Union unionRel = call.rel(0);
+    final DruidRel<?> firstDruidRel = call.rel(1);
+    final DruidRel<?> secondDruidRel = call.rel(2);
+
+    return isCompatible(unionRel, firstDruidRel, secondDruidRel, plannerContext);
+  }
+
+  @Override
+  public void onMatch(final RelOptRuleCall call)
+  {
+    final Union unionRel = call.rel(0);
+    final DruidRel<?> firstDruidRel = call.rel(1);
+    final DruidRel<?> secondDruidRel = call.rel(2);
+
+    if (firstDruidRel instanceof DruidFreeUnionDataSourceRel) {
+      // Unwrap and flatten the inputs to the Union.
+      final RelNode newUnionRel = call.builder()
+                                      .pushAll(firstDruidRel.getInputs())
+                                      .push(secondDruidRel)
+                                      .union(true, firstDruidRel.getInputs().size() + 1)
+                                      .build();
+
+      call.transformTo(
+          DruidFreeUnionDataSourceRel.create(
+              (Union) newUnionRel,
+              getColumnNames(firstDruidRel, plannerContext).get(),
+              firstDruidRel.getPlannerContext()
+          )
+      );
+    } else {
+      call.transformTo(
+          DruidFreeUnionDataSourceRel.create(
+              unionRel,
+              getColumnNames(firstDruidRel, plannerContext).get(),
+              firstDruidRel.getPlannerContext()
+          )
+      );
+    }
+  }
+
+  // Can only do UNION ALL of inputs that have compatible schemas (or schema mappings) and right side
+  // is a simple table scan
+  public static boolean isCompatible(
+      final Union unionRel,
+      final DruidRel<?> first,
+      final DruidRel<?> second,
+      @Nullable PlannerContext plannerContext
+  )
+  {
+    if (second instanceof DruidFreeUnionDataSourceRel) {
+      return false;
+    }
+
+    if (!unionRel.all && null != plannerContext) {
+      plannerContext.setPlanningError("SQL requires 'UNION' but only 'UNION ALL' is supported.");
+    }
+    return unionRel.all && isUnionCompatible(first, second, plannerContext);
+  }
+
+  /**
+   * For the current implementation the union is compatible iff the column's data types column and names are same
+   */
+  private static boolean isUnionCompatible(
+      final DruidRel<?> first,
+      final DruidRel<?> second,
+      @Nullable PlannerContext plannerContext
+  )
+  {
+    final Optional<List<String>> firstColumnNames = getColumnNames(first, plannerContext);
+    final Optional<List<String>> secondColumnNames = getColumnNames(second, plannerContext);
+    if (!firstColumnNames.isPresent() || !secondColumnNames.isPresent()) {
+      // No need to set the planning error here
+      return false;
+    }
+    if (!firstColumnNames.equals(secondColumnNames)) {
```

Comment:
Due to casting Calcite might change the name to something like EXPR$0. In such a case this does not allow the onMatch to trigger. What's our plan of action for handling such cases ?

Reply:
I am debating whether to act stringently and only allow the same names and types (no implicit casts) when using union all with this rule as well. Is there a way to realize that there is an implicit cast done, if so, then we can build something around it to remap it to the original variable name, otherwise it would be a hassle for the user since he won't be able to reference the original column.

Sidenote: I am debating whether to pull the planning changes out of the PR into their separate PR.

Reply:
If the union is nested inside a subquery, it would be difficult for the upper callers to reference it once it has changed into the form EXPR$0, due to implicit casting. Therefore if we can identify that it has been cast implicitly, then we should remap it back to the original column name, else I am debating that we should include name and the type check,